### PR TITLE
WIP: SOSS-ROS2 with xtypes

### PR DESCRIPTION
**Pull request by franava**

### CODE REVIEW

File: packages/ros2/src/MetaPublisher.cpp [hunk 1]
```diff
@@ -30,32 +29,32 @@ class MetaPublisher : public soss::TopicPublisher
 public:
 
   MetaPublisher(
-      StringTemplate&& topic_template,
+		    const std::string& topic_name,
```
Comment (lemunozm) [inline packages/ros2/src/MetaPublisher.cpp]: Please, be careful with the tabs!

File: packages/ros2/src/SystemHandle.cpp [hunk 2]
```diff
@@ -262,30 +242,32 @@ std::shared_ptr<TopicPublisher> SystemHandle::advertise(
...
+
   return true;
 }
 
 //==============================================================================
 std::shared_ptr<ServiceProvider> SystemHandle::create_service_proxy(
-    const std::string& service_name,
-    const std::string& service_type,
-    const YAML::Node& configuration)
+      const std::string& service_name,
```
Comment (lemunozm) [inline packages/ros2/src/SystemHandle.cpp]: Soss project uses a different indentation that our project.

File: packages/ros2/src/SystemHandle.cpp [hunk 3]
```diff
@@ -128,18 +130,17 @@ bool SystemHandle::configure(
...
-
-    success += set_platform_env("ROS_DOMAIN_ID", domain.c_str(), true);
+    success += setenv("ROS_DOMAIN_ID", domain.c_str(), true);
 
     _node = std::make_shared<rclcpp::Node>(name, ns);
 
     if(previous_domain.empty())
     {
-      success += unset_platform_env("ROS_DOMAIN_ID");
+        success += unsetenv("ROS_DOMAIN_ID");
```
Comment (lemunozm) [inline packages/ros2/src/SystemHandle.cpp]: Why this change?

File: packages/ros2/resources/convert__srv.cpp.em [hunk 4]
```diff
@@ -72,23 +72,28 @@ const std::string g_response_name = g_srv_name + ":response";
 namespace {
 
 //==============================================================================
-soss::Message initialize_request()
+xtypes::DynamicData&& initialize_request()
```
Comment (lemunozm) [inline packages/ros2/resources/convert__srv.cpp.em]: Why as rvalue? The copy elision should be enoguh

File: packages/ros2/resources/convert__msg.hpp.em [hunk 5]
```diff
@@ -74,43 +77,49 @@ using Ros2_Msg = @(cpp_msg_type);
 const std::string g_msg_name = "@(msg_type_string)";
 
 //==============================================================================
-inline soss::Message initialize()
+inline std::shared_ptr<xtypes::StructType> initialize()
```
Comment (lemunozm) [inline packages/ros2/resources/convert__msg.hpp.em]: Use Dynamic::Ptr instead

File: packages/ros2/resources/convert__msg.cpp.em [hunk 6]
```diff
@@ -69,16 +69,17 @@ private:
...
+    _callback(dynData);
   }
 
   // Save the SOSS callback that we were given by the soss-ros2 plugin
   TopicSubscriberSystem::SubscriptionCallback _callback;
 
   // Save a pre-initialized copy of the message so that we don't need to
   // allocate and deallocate more than necessary
-  soss::Message _message;
+  std::shared_ptr<xtypes::StructType> _dynamicTypeP;
```
Comment (lemunozm) [inline packages/ros2/resources/convert__msg.cpp.em]: use snake_case as the rest of the code.

File: packages/ros2/resources/convert__msg.cpp.em [hunk 7]
```diff
@@ -27,7 +27,7 @@ conversion_dependency = 'soss/rosidl/ros2/{}/msg/convert__msg__{}.hpp'.format(
 
 // Include the API header for this message type
 #include <@(conversion_dependency)>
-
+#include <xtypes/xtypes.hpp>
```
Comment (lemunozm) [inline packages/ros2/resources/convert__msg.cpp.em]: You should use soss/message from soss.

File: packages/ros2/resources/convert__srv.cpp.em [hunk 8]
```diff
@@ -112,23 +118,28 @@ void request_to_soss(const Ros2_Request& from, soss::Message& to)
...
+
 @[for field in alphabetical_response_fields]@
-  soss::Convert<Ros2_Response::_@(field.name)_type>::add_field(msg, "@(field.name)");
+//    type = evalType(field) ;
+//    msg.add_member("@(field.name)", type);
 @[end for]@
 
-  return msg;
+  DynamicData msg(type);
+  return std::move(msg);
```
Comment (lemunozm) [inline packages/ros2/resources/convert__srv.cpp.em]: This must no compile

File: packages/ros2/src/SystemHandle.cpp [hunk 9]
```diff
@@ -223,11 +224,12 @@ SystemHandle::~SystemHandle()
 
 //==============================================================================
 bool SystemHandle::subscribe(
-    const std::string& topic_name,
-    const std::string& message_type,
-    SubscriptionCallback callback,
-    const YAML::Node& configuration)
+      const std::string& topic_name,
```
Comment (lemunozm) [inline packages/ros2/src/SystemHandle.cpp]: Careful with the indentations, soss repo uses its own indentation.

File: packages/ros2/src/SystemHandle.hpp [hunk 10]
```diff
@@ -53,29 +54,29 @@ class SystemHandle : public virtual FullSystem
   ~SystemHandle() override;
 
   // Documentation inherited
-  bool subscribe(
+		bool subscribe(
```
Comment (lemunozm) [inline packages/ros2/src/SystemHandle.hpp]: Tabs!

File: packages/ros2/resources/convert__msg.cpp.em (hunk 7 above)
Comment (franava) [reply packages/ros2/resources/convert__msg.cpp.em]: I changed it, yet the PR tells me it is still on xtypes...

File: packages/ros2/resources/convert__srv.cpp.em (hunk 8 above)
Comment (franava) [reply packages/ros2/resources/convert__srv.cpp.em]: it did not, has been changed

File: packages/ros2/src/SystemHandle.cpp (hunk 9 above)
Comment (franava) [reply packages/ros2/src/SystemHandle.cpp]: this also has been changed but still looks like it has not.

File: packages/ros2/src/SystemHandle.hpp (hunk 10 above)
Comment (franava) [reply packages/ros2/src/SystemHandle.hpp]: same as the above. could you give it a look?

File: packages/ros2/resources/convert__msg.hpp.em (hunk 5 above)
Comment (franava) [reply packages/ros2/resources/convert__msg.hpp.em]: changed to DynamicData.

File: packages/ros2/resources/convert__srv.cpp.em (hunk 4 above)
Comment (franava) [reply packages/ros2/resources/convert__srv.cpp.em]: this also has been changed, no longer rvalue

File: packages/ros2/src/SystemHandle.cpp (hunk 3 above)
Comment (franava) [reply packages/ros2/src/SystemHandle.cpp]: I sincerely cannot track it back... I just rolled this change back
